nbcelltests/tests/test_test.py

```diff
@@ -60,28 +60,111 @@ def _import_from_path(path, module_name):
+    tf_name = tf.name
+    try:
+        # the module name (__name__) doesn't really matter, but
+        # will be nbcelltests.tests.test_tests.X, where X is
+        # whatever concrete subclass is this method belongs to.
+        generated_module = _import_from_path(run(notebook, filename=tf_name), module_name)
+    finally:
+        try:
+            tf.close()
+        except Exception:
+            pass
+        os.remove(tf_name)
+
+    return generated_module
+
+
 class _TestCellTests(unittest.TestCase):
     # abstract :)
 
     @classmethod
     def setUpClass(cls):
-        """
-        Generate test file from notebook, then import it, and make the
-        resulting module available as "generated_tests" class attribute.
-        """
-        tf = tempfile.NamedTemporaryFile(mode='w', suffix='.py', delete=False, encoding='utf8')
-        tf_name = tf.name
+        assert hasattr(cls, "NBNAME"), "Subclasses must have NBNAME attribute."  # TODO: make actually abstract
+        cls.generated_tests = _generate_test_module(notebook=cls.NBNAME, module_name="nbcelltests.tests.%s.%s" % (__name__, cls.__name__))
+
+    def _assert_skipped(self, mthd, reason):
+        # TODO: actually calling the skipped method here ends
+        # everything! It's some pytest issue, I think. So for now we
+        # are checking that it would be skipped, rather than that it is
+        # definitely skipped.
+        msg = "Should have generated a skipped test method"
+        self.assertTrue(hasattr(mthd, '__unittest_skip__'), msg=msg)
+        self.assertEqual(mthd.__unittest_skip__, True, msg=msg)
+        self.assertEqual(mthd.__unittest_skip_why__, reason, msg="Skip reason should have been %s" % reason)
+
+
+class TestMethodGenerationError(_TestCellTests):
+    """Tests of things that should fail during test script generation"""
+
+    NBNAME = NONCODE
+
+    @classmethod
+    def setUpClass(cls):
+        # we're testing what happens in setUpClass
+        pass
+
+    def test_non_code_cell_with_test_causes_error(self):
         try:
-            # the module name (__name__) doesn't really matter, but
-            # will be nbcelltests.tests.test_tests.X, where X is
-            # whatever concrete subclass is this method belongs to.
-            cls.generated_tests = _import_from_path(path=run(cls.NBNAME, filename=tf_name), module_name="nbcelltests.tests.%s.%s" % (__name__, cls.__name__))
-            tf.close()
-        finally:
-            os.remove(tf_name)
+            _generate_test_module(self.NBNAME, "module.name.irrelevant")
+        except ValueError as e:
+            assert e.args[0] == 'Cell 1 is not a code cell, but metadata contains test code!'
+        else:
+            raise Exception("Test script should fail to generate")
+
+
+class TestSkips(_TestCellTests):
+    """Tests that various conditions result in skipped tests"""
+
+    NBNAME = SKIPS
+
+    # the tests are independent, so it's fine to call setUpClass and
```

Comment:
It is fine to call it, but we might be missing some conditions that are only triggered if it is not. Ok since you mention a way to resolve in the future 👍

Reply:
This can be improved when we generate test notebooks. But ideally we wouldn't be invoking the setup and test methods manually in the first place, I guess.